### PR TITLE
Support ref field assignment in object initializers

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/62120
Corresponding spec update: https://github.com/dotnet/csharplang/pull/6277
Relates to test plan https://github.com/dotnet/roslyn/issues/59194

### CODE REVIEW

**Comment by Youssef1313** [comment]
> Fixes #62578

Should be #62120?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -12753,23 +12753,32 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
...
+                // Name = ref <expr>
+                // Name =  <expr>
                 isObjectInitializer = true;
                 return this.ParseObjectInitializerNamedAssignment();
             }
             else
             {
-                return this.ParseExpressionCore();
+                // <expr>
+                // ref <expr>
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:12780]: :memo: This case and the dictionary initializer case above remain errors in binding, but the parser now tolerates `ref` (more graceful failure).

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -1573,6 +1581,25 @@ private BoundExpression InferTypeForDiscardAssignment(BoundDiscardExpression op1
...
+
+            static string getName(BoundExpression expr)
+            {
+                if (expr.ExpressionSymbol is { Name: var name })
+                {
+                    return name;
+                }
+                if (expr is BoundArrayAccess)
+                {
+                    return "array access";
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:1593]: >"array access" [](http://example.com/codeflow?start=27&length=14)

Do these strings need to be localized?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -4866,10 +4877,10 @@ private BoundExpression MakeBadExpressionForObjectCreation(SyntaxNode node, Type
 
                 case BoundKind.ArrayAccess:
                 case BoundKind.PointerElementAccess:
-                    return boundMember;
+                    return CheckValue(boundMember, valueKind, diagnostics);
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4880]: Right. We were returning so bypassing the `CheckValueKind` check below.
Without this fix, we'd have no error in `RefInitializer_OnArray` or `RefInitializer_OnPointer`

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 4]
```diff
@@ -11743,5 +11743,1261 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+var r = new R() { field = ref x };
+
+ref struct R
+{
+    public ref readonly int field;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:11985]: Added `RefInitializer_AssignInParameter`

**Comment by cston** [comment]
>             comp = CreateCompilation(sourceB, references: new[] { refA }, parseOptions: TestOptions.Regular11);

I think this `parseOptions:` should be kept, to ensure we're not reporting an error for `-langversion:11` specifically, not just the latest version.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:152 in 6980e20. [](commit_id = 6980e208f70bd74ee86eaf0fa9fbfbbb9e7b5c03, deletion_comment = True)

**Comment by jcouv** [comment]
> I think this parseOptions: should be kept, to ensure we're not reporting an error for -langversion:11 specifically, not just the latest version.

Yes, I've reverted that line just after pushing. It was over-eager search & replace

**Review by cston — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 5]
```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+                // (3,31): error CS0165: Use of unassigned local variable 'x'
+                // var r = new R() { field = ref x };
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x").WithArguments("x").WithLocation(3, 31)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_RHSTypeMustMatch()
+        {
+            // The right operand must be an expression that yields an lvalue designating a value of the same type as the left operand.
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:11880]: consider testing a case where there is an implicit conversion from the RHS type to the LHS type.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 6]
```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+}
+
+ref struct R2
+{
+    public ref int _f;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (7,31): error CS8331: Cannot assign to variable 'in int' because it is a readonly variable
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12040]: ideally, the diagnostic message would say `in int i` here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 7]
```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+    public event System.Action a;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (3,17): error CS0070: The event 'C.a' can only appear on the left hand side of += or -= (except when used from within the type 'C')
+                // var r = new C { a = ref x }; // 1
+                Diagnostic(ErrorCode.ERR_BadEventUsage, "a").WithArguments("C.a", "C").WithLocation(3, 17),
+                // (6,3): error CS0070: The event 'C.a' can only appear on the left hand side of += or -= (except when used from within the type 'C')
+                // c.a = ref x; // 2
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12461]: might be nice to verify the behavior here on e.g. `void M() { this.a = ref x; }` or `void M() { this.a += ref x; }`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 8]
```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+ref struct R<T>
+{
+    public ref T F;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (3,34): error CS8173: The expression must be of type 'dynamic' because it is being assigned by reference
+                // var r = new R<dynamic> { F = ref i }; // 1
+                Diagnostic(ErrorCode.ERR_RefAssignmentMustHaveIdentityConversion, "i").WithArguments("dynamic").WithLocation(3, 34)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12662]: consider including a success case here (e.g. `dynamic i = 42`).

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 9]
```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "field = ref x").WithArguments("x").WithLocation(20, 25),
+                // (29,25): error CS8168: Cannot return local 'x' by reference because it is not a ref local
+                //             r = new R { field = ref x }; // 3
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "field = ref x").WithArguments("x").WithLocation(29, 25),
+                // (36,21): error CS8168: Cannot return local 'x' by reference because it is not a ref local
+                //         r = new R { field = ref x }; // 4
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "field = ref x").WithArguments("x").WithLocation(36, 21)
+                );
+
+            source = @"
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12809]: consider makign this a separate test and/or including a comment which indicates what the difference is between the two scenarios.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 10]
```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+        {
+            var source = @"
+class C
+{
+    public static Container M3()
+    {
+        Container r = default;
+        {
+            int x = 42;
+            var r = new Container { item = { field = ref x } }; // 1
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12891]: consider including some success cases.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RefFieldParsingTests.cs [hunk 11]
```diff
@@ -414,17 +415,60 @@ public void ReadOnlyRefParameter(LanguageVersion languageVersion)
...
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+            }
+            EOF();
+        }
+
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(LanguageVersionFacts.CSharpNext)]
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RefFieldParsingTests.cs:460]: nit: I expected use of `LanguageVersion.CSharp11` here instead.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 6 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12040]: https://github.com/dotnet/roslyn/issues/62096

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 12]
```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
...
+}
+
+ref struct R
+{
+    public ref int field;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,39): error CS8331: Cannot assign to method 'C.Value()' because it is a readonly variable
```
[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12009]: [https://github.com/dotnet/roslyn/issues/62756](https://github.com/dotnet/roslyn/issues/62756)

**Review by 333fred — APPROVED** [review]
LGTM (commit 6)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 13]
```diff
@@ -12663,6 +12730,22 @@ public void RefInitializer_DynamicField()
...
+dynamic i = 42;
+var r = new R<dynamic> { F = ref i };
+
+ref struct R<T>
+{
+    public ref T F;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12746]: Does the generated code run successfully? Consider running the code.

**Review by cston — APPROVED** [review]

**Pull request merged**